Add joypad motion sensors tutorial (#11692)
godotengine/godot-docs: tutorials/inputs/controller_features.rst

        blink_tween.tween_interval(0.2)
        blink_tween.set_loops(3)

Motion sensors (accelerometer and gyroscope)
---------

There are several controller brands that have implemented accelerometer and gyroscope sensors
into their modern controllers, the biggest two of them being PlayStation and Nintendo.

Before we dive deeper into how motion sensors can be used in Godot games,
first we need to shortly explain how they work and what they do.

**Accelerometer** is a type of sensor that detects a device's acceleration in m/s²,
which, in simpler terms, means it can detect if the device (in our case, a controller) has been physically moved quickly.

> For example, it can detect if the player quickly raises their controller, quickly moves it to the side

Reviewer: ```suggestion
For example, it can detect if the player quickly raises their controller, quickly moves it to the side,
```

Notice that we're using gyroscope calibration here by calling :ref:`Input.start_joy_motion_sensors_calibration()<class_Input_method_start_joy_motion_sensors_calibration>`
and :ref:`Input.stop_joy_motion_sensors_calibration()<class_Input_method_stop_joy_motion_sensors_calibration>`.
That's because modern gyroscopes often need calibration. This is like how a `weighing scale <https://en.wikipedia.org/wiki/Weighing_scale>`__ can need calibration to tell it what 'zero' is.
Like a weighing scale, a correctly calibrated gyroscope will give an accurate reading.
If you're using the gyro input as a mouse, which is the simplest application of a controller's gyroscope, you can find essential reading on `GyroWiki here <http://gyrowiki.jibbsmart.com/blog:good-gyro-controls-part-1:the-gyro-is-a-mouse>`__.
Calibration just means having the controller sit still and remembering the average reported angular velocity in each axis. This is the gyroscope's "bias".

After the controller's gyroscope has been enabled and correctly calibrated,
you can read its reported values by using :ref:`Input.get_joy_gyroscope()<class_Input_method_get_joy_gyroscope>`.

> Now we can move on to using the controller's accelerometer sensor in games.

Reviewer: I'd add two separate subheadings for accelerometer and gyro to make navigation easier

As mentioned earlier, accelerometer is a type of sensor that detects a device's acceleration in m/s²,
but the reported values also include **the force of gravity**,
and to get just the acceleration we can subtract the gravity values from the accelerometer values
like this: ``Input.get_joy_accelerometer(device) - Input.get_joy_gravity(device)``.

Note that due to how accelerometers work physically, when it's being quickly moved to the left
the reported X value is **positive**, and it's **negative** when the controller is moved to the right.

> Same can be said about other axes:

Reviewer: ```suggestion
The same is true for the other axes:
```

Notice that we're using gyroscope calibration here by calling :ref:`Input.start_joy_motion_sensors_calibration()<class_Input_method_start_joy_motion_sensors_calibration>`
and :ref:`Input.stop_joy_motion_sensors_calibration()<class_Input_method_stop_joy_motion_sensors_calibration>`.
That's because modern gyroscopes often need calibration. This is like how a weighing scale can need calibration to tell it what "zero" is.
Like a weighing scale, only a correctly calibrated gyroscope will give an accurate reading.
If you're using the gyro input as a mouse, which is the simplest application of a controller's gyroscope,
you can find essential reading on `GyroWiki here <http://gyrowiki.jibbsmart.com/blog:good-gyro-controls-part-1:the-gyro-is-a-mouse>`__.
Calibration just means having the controller sit still and remembering the average reported angular velocity on each axis.
This is the gyroscope's "bias".

After the controller's gyroscope has been enabled and correctly calibrated,
you can read its reported values by using :ref:`Input.get_joy_gyroscope()<class_Input_method_get_joy_gyroscope>`.

Accelerometer
~~~~~~~~~~~~~

> **Accelerometer** is a type of sensor that detects a device's acceleration in m/s².

Reviewer 2: ```suggestion
An **accelerometer** is a type of sensor that detects a controller's acceleration in m/s².
```

In simpler terms, this means it can detect if the device (in our case, a controller) has been physically moved quickly.
For example, it can detect if the player quickly raises their controller, quickly moves it to the side,

> or if the player is shaking the controller.

Reviewer 2: ```suggestion
For example, it can detect if the player quickly raises their controller, moves it to the side,
or shakes it.
```

Note that, unfortunately, it does not **not** mean that the device will be able to detect precise movements that can be used to replicate a VR controller's movement (for example).

Now we can move on to using the controller's accelerometer sensor in games.
As mentioned earlier, an accelerometer is a type of sensor that detects a device's acceleration in m/s²,
but the reported values also include **the force of gravity**,
and to get just the acceleration we can subtract the gravity values from the accelerometer values

> like this: ``Input.get_joy_accelerometer(device) - Input.get_joy_gravity(device)``.

Reviewer 2: ```suggestion
The acceleration that an accelerometer detects includes gravity by default.
To get *only* the acceleration imparted by the user, subtract gravity from the detected acceleration:

.. code-block::

    Input.get_joy_accelerometer(device) - Input.get_joy_gravity(device)

```

Note that due to how accelerometers work physically, when it's being quickly moved to the left
the reported X value is **positive**, and it's **negative** when the controller is moved to the right.
The same is true for the other axes:
when you move the controller upwards, the acceleration's Y value becomes negative,
when you move the controller downwards, the acceleration's Y value becomes positive,
when you quickly move the controller towards you its Z value becomes negative,

> and when you quickly move the controller away from you its Z value becomes positive.

Author: Thank you for your work! I will modify the PR later today to include your gamepad diagram! :)
The phone diagram will be useful too! See also https://github.com/godotengine/godot/pull/113635

> and when you quickly move the controller away from you its Z value becomes positive.

Author: Oh I just noticed the diagrams have text 😅 I think making the text a bit more readable would be great!
If gray text won't end up looking well enough, maybe we can also make it black with a white stroke?
May I also ask if you can invert the signs of the axes values? When you quickly move a controller to the left, the X axis value will be positive, for example.

> And due to how accelerometers work physically, after reporting movement in one direction

Reviewer 2: ```suggestion
Due to how accelerometers work physically, after movement in one direction stops
```

        if Input.has_joy_motion_sensors(0):
            accelerometer_example()

    func accelerometer_example():
        if not detect_accelerometer:
            return
            
        var acceleration = Input.get_joy_accelerometer(0) - Input.get_joy_gravity(0)
        if acceleration.length() > 10:
            if acceleration.x > 10:
                print("Moved left")
            elif acceleration.x < -10:
                print("Moved right") 
            if acceleration.y < -10:
                print("Moved up")
            elif acceleration.y > 10:
                print("Moved down")
            if acceleration.z < -10:
                print("Moved closer to the player")
            elif acceleration.z > 10:
                print("Moved away from the player")
                
            # After detecting movement in one direction, the accelerometer sensor
            # will briefly report movement in the opposite direction, even though the controller only moved once.
            # So we need to ignore these reported values for a short amount of time.
            detect_accelerometer = false
            await get_tree().create_timer(0.5, false).timeout

>             detect_accelerometer = true

Reviewer 2: ```suggestion
.. code-block::

    var detect_accelerometer = true
    
    # Change to make the game detect movement at different thresholds.
    # With a lower value, smaller movements will be detected, and with a
    # larger value, only big movements will be detected.
    const THRESHOLD = 10.0
    
    func _ready():
        # In this example we only use the first connected joypad (id 0).
        if 0 not in Input.get_connected_joypads():
            return

        if not Input.has_joy_motion_sensors(0):
            return

        # We must enable the motion sensors before using them.
        Input.set_joy_motion_sensors_enabled(0, true)

    func _process(delta):
        if Input.has_joy_motion_sensors(0):
            accelerometer_example()

    func accelerometer_example():
        if not detect_accelerometer:
            return
            
        var acceleration = Input.get_joy_accelerometer(0) - Input.get_joy_gravity(0)
        if acceleration.length() > THRESHOLD:
            if acceleration.x > THRESHOLD:
                print("Moved left")
            elif acceleration.x < -THRESHOLD:
                print("Moved right") 
            if acceleration.y < -THRESHOLD:
                print("Moved up")
            elif acceleration.y > THRESHOLD:
                print("Moved down")
            if acceleration.z < -THRESHOLD:
                print("Moved closer to the player")
            elif acceleration.z > THRESHOLD:
                print("Moved away from the player")
                
            # After detecting movement in one direction, the accelerometer sensor
            # will briefly report movement in the opposite direction, even though the controller only moved once.
            # So we need to ignore these reported values for a short amount of time.
            detect_accelerometer = false
            await get_tree().create_timer(0.5, false).timeout
            detect_accelerometer = true
```

Move the threshold value into a constant, so it's a bit easier for people to modify

        blink_tween.tween_interval(0.2)
        blink_tween.set_loops(3)

Motion sensors (gyroscope and accelerometer)
--------------------------------------------

There are several controller brands that have implemented gyroscope and accelerometer sensors
into their modern controllers, the biggest two being PlayStation and Nintendo.

To check if a connected controller has motion sensors, we can use :ref:`Input.has_joy_motion_sensors()<class_Input_method_has_joy_motion_sensors>`.

To use the controller motion sensors, we first have to enable them, because they are disabled by default
so they don't drain the controller's battery faster in games that don't use those features.
We can enable them by calling :ref:`Input.set_joy_motion_sensors_enabled()<class_Input_method_set_joy_motion_sensors_enabled>`.

Gyroscope
~~~~~~~~~

A **gyroscope** is a type of sensor that detects the controller's rotation.
They can be useful to make a camera that rotates as the player tilts their controller.
Here are some notable examples:

- In *Helldivers 2, Horizon Forbidden West, Star Wars: Dark Forces Remaster*, the controller's gyroscope is used for camera rotation ("gyro aiming").

> - In *Death Stranding*, the controller's gyroscope can be used for Controls for Soothing BB.

Author: Hm, I might need to take a look on how exactly BB soothing works with gyro and modify the text accordingly.

That's because modern gyroscopes often need calibration. This is like how a weighing scale can need calibration to tell it what "zero" is.
Like a weighing scale, only a correctly calibrated gyroscope will give an accurate reading.
During calibration, the user sets the controller down on a flat surface. The controller then determines what values its gyroscope reports when it is actually not moving at all (its "bias"), and uses this information to make its rotation data more accurate.

If you're using the gyroscope input as a mouse, which is the simplest application of a controller's gyroscope,
you can find essential reading on `GyroWiki here <http://gyrowiki.jibbsmart.com/blog:good-gyro-controls-part-1:the-gyro-is-a-mouse>`__.

After the controller's gyroscope has been enabled and correctly calibrated,
you can read its reported values by using :ref:`Input.get_joy_gyroscope()<class_Input_method_get_joy_gyroscope>`.

Accelerometer
~~~~~~~~~~~~~

An **accelerometer** is a type of sensor that detects a controller's acceleration in m/s².
For example, it can detect if the player quickly raises their controller, moves it to the side, or shakes it.

> .. warning::

Reviewer: ```suggestion

.. warning::
```

>     Note that, unfortunately, it does not **not** mean that the device will be able to detect precise movements that can be used to replicate a VR controller's movement (for example).

Reviewer: ```suggestion
    Note that, unfortunately, it does not **not** mean that the device will be able to detect
    precise movements that can be used to replicate a VR controller's movement (for example).
```

        blink_tween.tween_interval(0.2)
        blink_tween.set_loops(3)

Motion sensors (gyroscope and accelerometer)
--------------------------------------------

There are several controller brands that have implemented gyroscope and accelerometer sensors
into their modern controllers, the biggest two being PlayStation and Nintendo.

To check if a connected controller has motion sensors, use :ref:`Input.has_joy_motion_sensors()<class_Input_method_has_joy_motion_sensors>`.

Motion sensors are disabled by default to avoid draining the controller battery when games don't use those features.
To enable them, call :ref:`Input.set_joy_motion_sensors_enabled()<class_Input_method_set_joy_motion_sensors_enabled>`.

Note that the axes of the values that controller's motion sensors report are always relative to the controller's natural orientation.

> Here's an image of the axes mapping for more clarity:

Author: Not sure, since we already have a text version of the axes explained for accelerometer and gyroscope below. 😅

> Here's an image of the axes mapping for more clarity:

Reviewer 2: In that case, I think it'd make sense to move that explanation up to here, so that the user sees it first alongside the image, before going into the uses of it later on.

.. image:: img/controller_axes.webp 

Gyroscope
~~~~~~~~~

A **gyroscope** is a type of sensor that detects the controller's rotation.
They can be useful to make a camera that rotates as the player tilts their controller.

> Here are some notable examples:

Reviewer 2: ```suggestion
Here are some notable examples of gyroscope use in games:
```

- In *Helldivers 2*, *Horizon Forbidden West*, and *Star Wars: Dark Forces Remaster*, the controller's gyroscope is used for camera rotation ("gyro aiming").
- In *Death Stranding*, BB can be soothed with controller's rotation, which is picked up by the gyroscope.

> - In *For Honor*, the controller's gyroscope is used to parry attacks.

Author: I know nothing about the game either, and I currently can't find a video or text that explains this mechanic, so I removed this game from the list here.

		var gyro := Input.get_joy_gyroscope(0)
		node.rotation.x -= -gyro.y * GYRO_SENSITIVITY * delta  # Use rotation around the Y axis (yaw) here.
		node.rotation.y += -gyro.x * GYRO_SENSITIVITY * delta  # Use rotation around the X axis (pitch) here.

Note that before using the gyroscope's data, we must first calibrate it by calling :ref:`Input.start_joy_motion_sensors_calibration()<class_Input_method_start_joy_motion_sensors_calibration>`
and :ref:`Input.stop_joy_motion_sensors_calibration()<class_Input_method_stop_joy_motion_sensors_calibration>`.
That's because modern gyroscopes often need calibration. This is like how a weighing scale can need calibration to tell it what "zero" is.
Like a weighing scale, only a correctly calibrated gyroscope will give an accurate reading.
During calibration, the user sets the controller down on a flat surface.
The controller then determines what values its gyroscope reports when it is actually not moving at all (its "bias"),
and uses this information to make its rotation data more accurate.

See `the article on GyroWiki here <http://gyrowiki.jibbsmart.com/blog:good-gyro-controls-part-1:the-gyro-is-a-mouse>`

> for information on how to use gyroscope input as a mouse.

Reviewer 3: ```suggestion
See `the article on GyroWiki <http://gyrowiki.jibbsmart.com/blog:good-gyro-controls-part-1:the-gyro-is-a-mouse>`__
for information on how to use gyroscope input as a mouse.
```

PS: That website triggers an HTTP-only warning on Firefox in strict mode. @JibbSmart Could you look into setting up HTTPS with e.g. Let's Encrypt? This would avoid the warning.
